File: Haskell/Implementations/Decorator/BackgroundColor.hs
module BackgroundColor
    ( BackgroundColor(..)
    , defaultColor
    ) where

import BaseColor (BaseColor(..))
import Color (Color(..))

newtype BackgroundColor = MkBackgroundColor BaseColor

instance Color BackgroundColor where
    colorCode (MkBackgroundColor c) = "\x001b[" <> cc <> "m"
      where
        cc = case c of
            Black     -> "40"
            Red       -> "41"
            Green     -> "42"
            Yellow    -> "43"
            Blue      -> "44"
            Purple    -> "45"
            LightBlue -> "46"
            White     -> "47"

defaultColor :: BackgroundColor
defaultColor = MkBackgroundColor Black

File: Java/Implementations/Decorator/BackgroundColor.java
public class BackgroundColor implements Color {
	private BaseColor color;

	public BackgroundColor(BaseColor color) {
		this.color = color;
	}

	private String cc() {
		switch (this.color) {
			case BLACK:     return "40";
			case RED:       return "41";
			case GREEN:     return "42";
			case YELLOW:    return "43";
			case BLUE:      return "44";
			case PURPLE:    return "45";
			case LIGHTBLUE: return "46";
			case WHITE:     return "47";
			default:   return "40";
		}
	}

	public String colorCode() {
		return "\u001b[" + this.cc() + "m";
	}

	public static Color DEFAULT = new BackgroundColor(BaseColor.BLACK);
}
